tests/jlm/util/TestTarjanSCC.cpp

```diff
@@ -0,0 +1,204 @@
+  {
+    return successors[node];
+  };
+
+  std::vector<size_t> sccIndex;
+  std::vector<size_t> topologicalOrder;
+  auto numSccs = jlm::util::FindStronglyConnectedComponents(
+      numNodes,
+      GetSuccessors,
+      sccIndex,
+      topologicalOrder);
+  ValidateTopologicalOrderAndSccIndices(numNodes, GetSuccessors, sccIndex, topologicalOrder);
+
+  assert(numSccs == 3);
+  // 5 has to be at the end
+  assert(sccIndex[5] == 0);
+  assert(topologicalOrder[numNodes - 1] == 5);
+  // 6 has to be at the beginning
+  assert(sccIndex[6] == 2);
+  assert(topologicalOrder[0] == 6);
+  // The rest belong to the middle SCC
+  for (size_t i = 0; i < 5; i++)
+    assert(sccIndex[i] == 1);
+}
+
+// Creates a chain of diamonds, possibly with a single back-edge
+// When there is a back-edge, there should be a single SCC with order > 1
+static void
+TestDiamondChain(bool withBackEdge)
+{
+  /* Should be 3x+1, where x+1 is the number of "knots", the bottlenecks in the diamonds
+     O   O   etc.
+    / \ / \ /
+   O   O   O   <- these are the knots
+    \ / \ / \
+     O   O   etc.
+  */
+  const size_t numNodes = 97;
+  std::vector<std::vector<size_t>> successors(numNodes);
+
+  // Use a permutation to "randomize" the node indices
+  std::vector<size_t> perm(numNodes);
+  for (size_t i = 0; i < numNodes; i++)
+  {
+    // 50 and 97 are coprime, so this will assign each node a unique new index
+    perm[i] = (i * 50) % numNodes;
+  }
+
+  for (size_t knot = 0; knot + 3 < numNodes; knot += 3)
+  {
+    successors[perm[knot]].push_back(perm[knot + 1]);
+    successors[perm[knot]].push_back(perm[knot + 2]);
+    successors[perm[knot + 1]].push_back(perm[knot + 3]);
+    successors[perm[knot + 2]].push_back(perm[knot + 3]);
+  }
+
+  if (withBackEdge)
+  {
+    // Leave 5 nodes on the end, and 5 at the beginning, out of the big loop
+    successors[perm[numNodes - 6]].push_back(perm[5]);
+  }
+
+  auto GetSuccessors = [&](size_t node)
+  {
+    return successors[node];
+  };
+
+  std::vector<size_t> sccIndex;
+  std::vector<size_t> topologicalOrder;
+  auto numSccs = jlm::util::FindStronglyConnectedComponents(
+      numNodes,
+      GetSuccessors,
+      sccIndex,
+      topologicalOrder);
+  ValidateTopologicalOrderAndSccIndices(numNodes, GetSuccessors, sccIndex, topologicalOrder);
+
+  if (withBackEdge)
+  {
+    // With a back edge, only 5 nodes at the beginning and 5 nodes and the end are not in the loop
+    assert(numSccs == 5 + 5 + 1);
+    // All the middle nodes belong to one huge SCC
+    auto loopScc = sccIndex[perm[5]];
+    for (size_t i = 6; i < numNodes - 5; i++)
+    {
+      assert(sccIndex[perm[i]] == loopScc);
+    }
+  }
+  else
+  {
+    // Without any back-edges, every node should be its own SCC
+    assert(numSccs == numNodes);
+  }
+}
+
+static int
+TestTarjanSCC()
+{
+  TestDAG();
+  TestCycles();
+  TestDiamondChain(false);
+  TestDiamondChain(true);
+  return 0;
+}
+
+JLM_UNIT_TEST_REGISTER("jlm/util/TestTarjanSCC", TestTarjanSCC)
```

Comment:
I know we have (mostly) not done it so far, but I would like to move away from the single test function per file and move to multiple tests per file such that we have the chance to invoke them individually. In other words, it should look like this:

```
static int
TestDag()
{
...
}
JLM_UNIT_TEST_REGISTER("jlm/util/TestTarjanSCC-TestDag", TestDAG)

static int
TestCycles()
{
...
}
JLM_UNIT_TEST_REGISTER("jlm/util/TestTarjanSCC-TestCycles", TestCycles)
...
```